### PR TITLE
Fix nested create, upsert and update on lists of inputs

### LINKED ISSUE #1315: Using @upsert or @update on an array of input types gives unexpected behavior
Copied over from https://github.com/nuwave/lighthouse/pull/1185#issuecomment-617762552

**Describe the bug**
I'm unable to use `@create` or `@upsert` on an array of input types.
I have the following types and mutation

```graphql
type AssessedRoom {
    id: ID
    name: String!
    picture: String
    assessment: Assessment @belongsTo
    room: Room @belongsTo
    answers: [Answer] @hasMany
}

input UpsertAssessedRoom {
    id: ID
    name: String
    picture: Upload @upload
    room: ID! @connect
    assessment: ID! @connect
    answers: [UpsertAnswer] @upsert
}

extend type Mutation @guard(with: ["api"]){
    upsertAssessedRoom(input: UpsertAssessedRoom! @spread): AssessedRoom @upsert
}
```

According to the documentation [here](https://lighthouse-php.com/4.11/concepts/arg-resolvers.html#solution) the `answers: [UpsertAnswer] @upsert` should work here inside the `UpsertAssessedRoom` input type. But what I found out is that it's not working.

Given the following test:

```php
 /** @test */
    public function when_the_answers_do_provide_the_value_based_on_the_question_type_no_validation_error_is_given()
    {
        $room = create_room(['is_picture_required' => false]);
        $assessment = create_assessment();
        $text = Text::make('How are you doing today?')->build();
        $boolean = Boolean::make('Are you ok?')->build();
        $multipleChoice = MultipleChoice::make('Star Wars or Star Trek?')->options(
            Option::make('Star Wars'),
            Option::make('Star Trek'),
        )->build();
        $option1 = $multipleChoice->options[0];

        $measurement = Measurement::make('How tall are you?')->inCentimeters()->max(200)->build();

        $mutation = $this->graphQL(/* @lang GraphQL */ '
        mutation ($input: UpsertAssessedRoom!){
          upsertAssessedRoom(input: $input){
            name
            room { id }
            assessment { id }
            answers {
              text
              bool
              picture
              measurement
              question {
                id
              }
              option {
                id
              }
            }
          }
        }
        ', [
            'input' => [
                'name'       => 'Zolder',
                'room'       => $room->id,
                'assessment' => $assessment->id,
                'answers'    => [
                    [
                        'question' => $text->id,
                        'text'     => 'good!'
                    ],
                    [
                        'question' => $boolean->id,
                        'bool'     => false,
                    ],
                    [
                        'question' => $multipleChoice->id,
                        'option'   => $option1->id,
                    ],
                    [
                        'question'    => $measurement->id,
                        'measurement' => 190
                    ],
                ],
            ],
        ]);


        $mutation->assertJson([
            'data' => [
                'upsertAssessedRoom' => [
                    'name'       => 'Zolder',
                    'room'       => ['id' => $room->id],
                    'assessment' => ['id' => $assessment->id],
                    'answers'    => [
                        [
                            'question' => ['id' => $text->id],
                            'text'     => 'good!'
                        ],
                        [
                            'question' => ['id' => $boolean->id],
                            'bool'     => false,
                        ],
                        [
                            'question' => ['id' => $multipleChoice->id],
                            'option'   => ['id' => $option1->id],
                        ],
                        [
                            'question'    => ['id' => $measurement->id],
                            'measurement' => 190,
                        ]
                    ]
                ]
            ]
        ]);
```

I get the following error:
```json
Unable to find JSON: 

[{
    "data": {
        "upsertAssessedRoom": {
            "name": "Zolder",
            "room": {
                "id": 11
            },
            "assessment": {
                "id": 11
            },
            "answers": [
                {
                    "question": {
                        "id": 38
                    },
                    "text": "good!"
                },
                {
                    "question": {
                        "id": 39
                    },
                    "bool": false
                },
                {
                    "question": {
                        "id": 40
                    },
                    "option": {
                        "id": 19
                    }
                },
                {
                    "question": {
                        "id": 41
                    },
                    "measurement": 190
                }
            ]
        }
    }
}]

within response JSON:

[{
    "data": {
        "upsertAssessedRoom": {
            "name": "Zolder",
            "room": {
                "id": "11"
            },
            "assessment": {
                "id": "11"
            },
            "answers": [
                {
                    "text": "good!",
                    "bool": false,
                    "picture": null,
                    "measurement": 190,
                    "question": {
                        "id": "41"
                    },
                    "option": {
                        "id": "19"
                    }
                }
            ]
        }
    }
}].
```

it's only making **one** answer instead of the expected four. But this one answer has  **all** the fields set to values of the all the different answers I try to make. 

Seems like somewhere the first relation that is created keeps being passed on when new instances should be created.

Just to note, `@create` gives the same result.

**Expected behavior/Solution**

I expect to be able to use `@create` and `@upsert` on an array of input types, making and/or updating the given models, instead of just creating one model.

**Environment**

Lighthouse Version: `^4.11`
Laravel Version: `v7.6.0`

### PR DESCRIPTION
**Pull request by megawubs**
- [x] Added or updated tests
- [ ] ~~Added Docs for all relevant versions~~
- [x] Updated CHANGELOG.md

Resolves #1315 

**Changes**

Pass a new instance of the model to the actual update method.

**Breaking changes**
None

### CODE REVIEW

**Comment by spawnia** [comment]
@megawubs the branch was quite out of date. There is a GitHub bot which automatically keeps your branches up to date: https://probot.github.io/apps/pull/ (Beware that it nukes any changes you made on the `master` branch of any fork)

**Comment by spawnia** [comment]
Concerning https://github.com/nuwave/lighthouse/pull/1316/commits/26a9406891e8e876e7f1ea5d66f0ae37a4478f94, we can not change that until we drop support for PHPUnit 6.

**Comment by megawubs** [comment]
@spawnia does this need an update on the docs?

**Comment by megawubs** [comment]
> Concerning [26a9406](https://github.com/nuwave/lighthouse/commit/26a9406891e8e876e7f1ea5d66f0ae37a4478f94), we can not change that until we drop support for PHPUnit 6.

Ah ok, i'll revert that then

File: tests/Integration/Schema/Directives/UpsertDirectiveTest.php [hunk 1]
```diff
@@ -16,27 +16,28 @@ public function testNestedArgResolver(): void
             'name' => 'old',
         ]);
 
-        $this->schema .= /** @lang GraphQL */ '
+        $this->schema .= /** @lang GraphQL */
```
Comment (megawubs) [inline tests/Integration/Schema/Directives/UpsertDirectiveTest.php:19]: I don't know why my phpstorm does this. Looking into it now

**Comment by spawnia** [comment]
> @spawnia does this need an update on the docs?

We actually have an example in the docs that is now fixed:

```graphql
input CreateTaskInput {
  name: String!
  notes: [CreateNoteInput!]
  notes: [CreateNoteInput!] @create
}
```

File: tests/Integration/Schema/Directives/CreateDirectiveTest.php [hunk 2]
```diff
@@ -77,7 +79,8 @@ public function testCanCreateFromInputObject(): void
 
     public function testCreatesAnEntryWithDatabaseDefaultsAndReturnsItImmediately(): void
     {
-        $this->schema .= /** @lang GraphQL */ '
+        $this->schema .= /** @lang GraphQL */
+            '
```
Comment (spawnia) [inline tests/Integration/Schema/Directives/CreateDirectiveTest.php:83]:
```suggestion
        $this->schema .= /** @lang GraphQL */ '
```

Same with the other changes like this.

File: tests/Integration/Schema/Directives/CreateDirectiveTest.php [hunk 3]
```diff
@@ -11,7 +11,7 @@ class CreateDirectiveTest extends DBTestCase
 {
     public function testCanCreateFromFieldArguments(): void
     {
-        $this->schema .= /** @lang GraphQL */ '
+        $this->schema .= /** @lang GraphQL */'
```
Comment (spawnia) [inline tests/Integration/Schema/Directives/CreateDirectiveTest.php]:
```suggestion
        $this->schema .= /** @lang GraphQL */ '
```

Comment (spawnia) [reply tests/Integration/Schema/Directives/CreateDirectiveTest.php]: I am changing it, nvm.

**Comment by megawubs** [comment]
> @megawubs the branch was quite out of date. There is a GitHub bot which automatically keeps your forks up to date: https://probot.github.io/apps/pull/ (Beware that it nukes any changes you made on the `master` branch of any fork)

🤔 I did a merge with upstream before I started. Never mind, looks like it still contains the fix.

**Comment by spawnia** [comment]
> I did a merge with upstream before I started. Never mind, looks like it still contains the fix.

AFAIK, doing that in PHPStorm does not automatically pull upstream?

**Comment by megawubs** [comment]
I think this is everything right?

**Comment by megawubs** [comment]
> > I did a merge with upstream before I started. Never mind, looks like it still contains the fix.
> 
> AFAIK, doing that in PHPStorm does not automatically pull upstream?

Did it through command line git.

**Comment by spawnia** [comment]
> I think this is everything right?

Yeah, great job! Thanks a lot for taking the time to write proper tests. I am going to release a new version right now.

**Comment by lorado** [comment]
Did I miss something? I thought there is no need to use `@create`/`@update` on nested mutations.

```graphql
        input CreateUserInput {
             name: String
             tasks: [CreateTaskInput!] @create
         }
```

It still works also without `@create` on `tasks` field, right? I don't understand the use case.

**Comment by spawnia** [comment]
@lorado without `@create`, you would have to add an intermediary `input` like this:

```graphql
input CreateUserInput {
    name: String
    tasks: [CreateTaskInput!]
}

input CreateTaskHasMany {
    create: [CreateTaskInput!]
}
```

You can skip that layer if you know that all you are ever going to do with that relation is specifically create.

**Pull request merged**